### ISSUE TITLE
Support for alternate base url

### ISSUE THREAD

**Issue by neonspectra** [dullage/flatnotes#183]
# Overview

Flatnotes expects to live at the root domain of the site at which it is hosted. This works great if you want to host it as a subdomain, but not great if you want to make it a subdirectory of a domain.

**IE:**
- ✅ `flatnotes.example.com`
- 🚫 `example.com/flatnotes`

This is a classic problem in web application hosting design, and there's a really good software-agnostic generalised discussion of the nature of the problem  [here](https://caddy.community/t/the-subfolder-problem-or-why-cant-i-reverse-proxy-my-app-into-a-subfolder/8575).

## Reproducing the Issue

All you need to do is host an instance of Flatnotes with your favourite reverse proxy in front of it. You will see that it works when using a domain with no baseurl redirection, but it breaks if you try to stuff Flatnotes into its own subpath.

This should be demonstrable with any reverse proxy of your choice. For the sake of demonstration, I've made some examples using Caddy for how to do this. I've used syntax that I am actively using for other services which support alternate baseurl configuration, so I'm confident that it should work if Flatnotes supported this in config.

### ✅ Working Config
Caddyfile snippet:
```
example.com {
  reverse_proxy flatnotes:8080
}
```
<img width="680" alt="Screenshot 2024-04-27 at 18 54 45" src="https://github.com/dullage/flatnotes/assets/10284121/342daab9-c7f5-4245-8044-e4ce6b76bacb">

### 🚫 Broken Config
Caddyfile snippet:
```
example.com {
redir /flatnotes /flatnotes/
handle /flatnotes/* {
  reverse_proxy flatnotes:8080
}
```
<img width="781" alt="Screenshot 2024-04-27 at 18 53 59" src="https://github.com/dullage/flatnotes/assets/10284121/d6f10af1-20d7-4c83-a0f1-78397e0aa4d2">

## Desired Functionality
Flatnotes should support an environment variables along the lines of `FLATNOTES_BASEURL` which accepts a path in the form of `/subdirectory`.

For example, some psudeocode:
- `FLATNOTES_BASEURL=/flatnotes` → ✅ `example.com/flatnotes`

### Why would you want to do this?
Sometimes you may not have a subdomain available in the DNS zone that you want to host flatnotes in. 

For example, if you are using [Tailscale](https://tailscale.com/) [Magic DNS](https://tailscale.com/kb/1081/magicdns) with your tailnet name, you only have one subdomain per active node. This means that if you want to have more than one service hosted on the same system, you need to host them as subdirectories rather than putting services on a given root domain.

**Comment by fawick** [comment]
Having a maintained way to set the base URL to a subfolder would be great, indeed.

First, I tried to come up with a solution by adding the `--root_path` option to the `uvicorn` call in `entrypoint.sh` but that was not enough. The browser tried to load the front-end assets without the subfolder.

Second, I added the `base` option to `vite.config.js`. That helped with the JS assets, but all requests to `/api/*` are still failing (e.g. `/api/config` or `/api/notes/[...]`).

**Comment by Code-Otto** [comment]
Also check out #49 (And PR #135), there's been a prior attempt at this but disregarded because of the added complexity

I'd appreciate the addition of such a feature since this is the last missing straw for proper Flatnotes integration in my home server

**Comment by fawick** [comment]
Thank you @Code-Otto for making me aware on the prior work. 

As it looks, #135 is probably rendered obsolete by the switch from Parcel to Vite (and #135 was rejected anyway).

**Comment by praisethemoon** [comment]
Hello! Also loving this project, and wanting this feature, keep it up! 🙌 nevertheless ❤️

**Comment by dullage** [comment]
Thank you very much for taking the time to provide the info here @kaliatech! The base element wasn't something I was aware of but solves a lot of the problems I was having.

I'm pleased to say that I now have a working build that supports a custom path prefix 🎉. I would appreciate some beta testers if anyone would like to give it a go. You just need to use the docker tag `dullage/flatnotes:custom-path-prefix` and set the `FLATNOTES_PATH_PREFIX` environment variable to something like `/flatnotes`.

⚠️Note: New image attachments will embed a URL that works even if the path prefix is changed. Unfortunately, existing attachment URLs will need to be manually updated (but only for existing users that want to use a path prefix). You'll need to change, for example, `/attachments/foo.png` to `attachments/foo.png`.

Duplicate Suggestion: #49
Relevant Commit: [a43b9cf](https://github.com/dullage/flatnotes/commit/a43b9cf592e9e88299b04bf252165a55d17641c8)
Previous PRs: #71 #135

**Comment by kaliatech** [comment]
Nice! It is working for me locally using a custom path and the `dullage/flatnotes:custom-path-prefix` image. 

Fwiw, I use [Caddy](https://caddyserver.com/) to reverse proxy to the docker container with a simple config like this:
```
    handle /custom-path {
      redir /custom-path /custom-path/
    }

    handle /custom-path* {
      reverse_proxy http://my-flatpath-server:8080
    }   
```

The only minor issue I noticed is that probably not all favicons are loading as expected. I'm seeing error like this in the dev console:

> GET https://www.mydomain.com/custom-path/assets/android-chrome-192x192.png 404 (Not Found)

Not sure if you intend to copy the favicons to the dist /assets folder (allowing users to override them) or had something else in mind. ..._if_ that is your intent, will probably want to eventually have instructions on how to override the root /favicon.ico as well.

Thanks for handling the root URL path issue so quickly!

**Comment by dullage** [comment]
Thanks for giving it a go @kaliatech and for reporting the issue with the PWA icons. This should be resolved in the latest build of the `:custom-path-prefix` image. If you get a chance, I'd appreciate it if you could pull the image and confirm you are no longer seeing the 404.

**Comment by kaliatech** [comment]
Looks good. I pulled latest and all favicons are now resolving without error.

I also noticed the container is reporting unhealthy status. So, I think you might need to apply the `FLATNOTES_PATH_PREFIX ` to healthcheck.

  - [Dockerfile#L61](https://github.com/dullage/flatnotes/blob/custom-root-path/Dockerfile#L61)
  - [healthcheck.sh#L3](https://github.com/dullage/flatnotes/blob/custom-root-path/healthcheck.sh#L3)

**Comment by dullage** [comment]
Thanks @kaliatech, another good catch. This has now been sorted, and a new image has been pushed.

_Note: Previously the app would add/remove leading/trailing slashes as necessary, but now the `FLATNOTES_PATH_PREFIX` value is validated to ensure it starts with but does not end with a slash (e.g. `/flatnotes`)._

**Comment by dullage** [comment]
Now available in 5.1.0.
